### PR TITLE
Restore find_unused_parameters for non-compile DDP training

### LINKED ISSUE #24675: Bug: Multi-GPU DDP training hangs at specific epoch on v8.4.60 but works perfectly on v8.4.9
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Multi-GPU

### Bug

Description

When training yolov8s using Multi-GPU DDP (2x RTX 3090), the training process completely hangs at a specific epoch without throwing any errors or exceptions.

    When epochs=120 and batch=120, it consistently hangs at Epoch 30.

    When epochs=30 and batch=120, it consistently hangs at Epoch 27.

    If I change batch=100, it runs smoothly to completion.

    Single-GPU training works completely fine under all configurations.

Crucial Version Clue:
This is a regression bug related to recent updates.

    ultralytics==8.4.9: Works perfectly (Runs to completion with batch=120 on dual GPUs).

Image

pip install -i https://pypi.tuna.tsinghua.edu.cn/simple ultralytics -U -q

    ultralytics==8.4.60: Reproducibly hangs as described above. Downgrading back to 8.4.9 immediately resolves the issue.

Image

### Environment

vision@vision-System-Product-Name /m/w/x/p/m/9/yolov8 (main) [1]> yolo check                                                                                                                                                                                        (pt29)
Ultralytics 8.4.60 🚀 Python-3.11.11 torch-2.9.1+cu128 CUDA:0 (NVIDIA GeForce RTX 3090, 24124MiB)
Setup complete ✅ (56 CPUs, 31.2 GB RAM, 24.5/152.6 GB disk)

OS                     Linux-6.8.0-111-generic-x86_64-with-glibc2.35
Environment            Linux
Python                 3.11.11
Install                pip
Path                   /home/vision/anaconda3/envs/pt29/lib/python3.11/site-packages/ultralytics
RAM                    31.20 GB
Disk                   24.5/152.6 GB
CPU                    Intel Xeon E5-2680 v4 2.40GHz
CPU count              56
GPU                    NVIDIA GeForce RTX 3090, 24124MiB
GPU count              2
CUDA                   12.8

numpy                  ✅ 2.3.4>=1.23.0
matplotlib             ✅ 3.10.7>=3.3.0
opencv-python          ✅ 4.11.0.86>=4.6.0
pillow                 ✅ 11.3.0>=7.1.2
pyyaml                 ✅ 6.0.3>=5.3.1
requests               ✅ 2.32.5>=2.23.0
scipy                  ✅ 1.16.3>=1.4.1
torch                  ✅ 2.9.1>=1.8.0
torch                  ✅ 2.9.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision            ✅ 0.24.1>=0.9.0
psutil                 ✅ 7.1.3>=5.8.0
polars                 ✅ 1.35.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18

### Minimal Reproducible Example

```
import sys
import os
from ultralytics import YOLO

if __name__ == "__main__":
    data_cfg_path = f"../cfg/SOD123-2D-RGB_v1_.yaml"
    model = YOLO(
        "../model/yolov8s.pt",
        task="detect",
    )

    epochs = 30  # 训练轮数

    model.train(
        data=data_cfg_path,  # 你的数据配置文件路径
        epochs=epochs,  # 训练轮数
        imgsz=640,  # 输入图像尺寸
        batch=120,
        name="train_" + dataset_name,  # 子目录名
        mosaic=0.8,  # 默认1，拼图增强的概率。 仅当允许图像中存在部分遮挡物体或多物体共存，且这些情况不会改变标签值时(类别)，才使用拼图增强。mosaic=1.0 表示所有图像都将进行拼图增强，
        workers=3,
        device=[0, 1],
    )
```

### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by raimbekovm**
## summary

multi-GPU DDP training could silently hang on datasets where some per-GPU sub-batches contain no objects, because `find_unused_parameters` was dropped from the DDP wrapper in #24413. this restores `find_unused_parameters=True` on the non-compile path while keeping `static_graph=True` under `torch.compile`, so params that receive no gradient on object-less batches no longer deadlock the collective. fixes #24675.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Restores `find_unused_parameters` for non-compile Distributed Data Parallel training to improve stability when some model parameters do not receive gradients on certain batches. ⚙️

### 📊 Key Changes
- Updates `DistributedDataParallel` setup in `ultralytics/engine/trainer.py`.
- Keeps `static_graph=True` only when `self.args.compile` is enabled.
- Adds `find_unused_parameters=True` when training without compile.
- Expands the inline comment to clarify the different roles of `static_graph` and `find_unused_parameters`.

### 🎯 Purpose & Impact
- Fixes a DDP training edge case where some parameters may be unused in specific batches, such as object-less batches. 🛠️
- Preserves the compile-specific optimization path while restoring safer behavior for standard non-compile distributed training.
- Especially helpful for workflows like pose training or other multi-branch losses where parameter usage can vary by batch. 📈

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @raimbekovm, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you shortly. Please review the checklist below to help ensure a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change looks clean and scoped appropriately. Restoring `find_unused_parameters` for non-compile DDP while preserving `static_graph=True` only for compile mode matches the stated intent and does not introduce any obvious correctness or maintainability issues in the diff shown.

**Comment by glenn-jocher** [comment]
Thanks for this, and for accurately diagnosing the regression. The same fix already landed on `main` via #24620 (commit `8b673ce5e`): `trainer.py` now sets `find_unused_parameters=not bool(self.args.compile)`, which is functionally identical to your change. Closing as already resolved — appreciate the contribution!
